Add a platform specific method to call when "gamepads recently accessed" state changes

beidson opened the pull request:
#### 01f11ddb6c104c73d97f7ba1a879b38be925e2c6
<pre>
Add a platform specific method to call when &quot;gamepads recently accessed&quot; state changes
<a href="https://rdar.apple.com/128171648">rdar://128171648</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=274238">https://bugs.webkit.org/show_bug.cgi?id=274238</a>

Reviewed by Aditya Keerthi.

Some Apple builds will implement this for now, and it can remain a no-op for others.

* Source/WebCore/platform/gamepad/cocoa/GameControllerSoftLink.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerSoftLink.mm:

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setGamepadsRecentlyAccessed:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::setGamepadsRecentlyAccessed):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::setGamepadsRecentlyAccessed):

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::recentGamepadAccessStateChanged):

Canonical link: <a href="https://commits.webkit.org/278887@main">https://commits.webkit.org/278887@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02f2aff72e28d2d7f54596fd0e4c6ac0af9c7471

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51818 "Failed to checkout and rebase branch from PR 28638") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31130 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4182 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55084 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/2510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37502 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2217 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/53917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28750 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/44702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23286 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1962 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56677 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/26939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44798 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11328 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/29073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1263,6 +1263,13 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+#if ENABLE(GAMEPAD)
+void WebPageProxy::platformGamepadsRecentlyAccessed(GamepadsRecentlyAccessed gamepadsRecentlyAccessed)
+{
+    [cocoaView().get() _setGamepadsRecentlyAccessed:(gamepadsRecentlyAccessed == GamepadsRecentlyAccessed::No) ? NO : YES];
```
Access to the underlying view is normally done via `PageClient`.

(also, `.get()` is unnecessary)

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3063,6 +3063,11 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
 #if ENABLE(GAMEPAD)
     void recentGamepadAccessStateChanged(PAL::HysteresisState);
+    enum class GamepadsRecentlyAccessed {
```
```suggestion
    enum class GamepadsRecentlyAccessed : bool {
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2047,9 +2047,15 @@ - (void)setMinimumViewportInset:(CocoaEdgeInsets)minimumViewportInset maximumVie
 }
 
-#if USE(APPLE_INTERNAL_SDK)
+#if USE(APPLE_INTERNAL_SDK) && __has_include(<WebKitAdditions/GameControllerSoftLinkAdditions.h>)
```
The `__has_include` doesn't seem quite right here, given this is a pre-existing import.

beidson replied:
It is not a preexisting import (I can point you to the additions patch as well if you'd like)

pxlcoder replied:
The import already exists on `main` though: https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm#L2051

beidson replied:
Apologies, I misunderstood - Without the context of the literal next line showing, I thought you were saying "<WebKitAdditions/GameControllerSoftLinkAdditions.h> is a preexisting import"

Yes, this was wrong. New patch coming.

beidson committed:
Add a platform specific method to call when "gamepads recently accessed" state changes
rdar://128171648
https://bugs.webkit.org/show_bug.cgi?id=274238

Reviewed by Aditya Keerthi.

Some Apple builds will implement this for now, and it can remain a no-op for others.

* Source/WebCore/platform/gamepad/cocoa/GameControllerSoftLink.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerSoftLink.mm:

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setGamepadsRecentlyAccessed:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::setGamepadsRecentlyAccessed):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::setGamepadsRecentlyAccessed):

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::recentGamepadAccessStateChanged):

Canonical link: https://commits.webkit.org/278887@main
